### ISSUE TITLE
Investigate novel pub/sub system test failures

### ISSUE THREAD

**Issue by dhermes** [GoogleCloudPlatform/gcloud-python#1085]
1.  `Subscription.pull(return_immedately=False, max_messages=2)` returns immediately (See #893)

1.  Missing ACK ID

    ```python
    ======================================================================
    ERROR: test_message_pull_mode_e2e (pubsub.TestPubsub)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/travis/build/GoogleCloudPlatform/gcloud-python/system_tests/pubsub.py", line 123, in test_message_pull_mode_e2e
        subscription.acknowledge(ack_ids)
      File "/home/travis/build/GoogleCloudPlatform/gcloud-python/.tox/system-tests/lib/python2.7/site-packages/gcloud/pubsub/subscription.py", line 230, in acknowledge
        method='POST', path='%s:acknowledge' % (self.path,), data=data)
      File "/home/travis/build/GoogleCloudPlatform/gcloud-python/.tox/system-tests/lib/python2.7/site-packages/gcloud/connection.py", line 419, in api_request
        error_info=method + ' ' + url)
    gcloud.exceptions.BadRequest: 400 You have not specified an ack ID in the request. (POST https://pubsub.googleapis.com/v1/projects/precise-truck-742/subscriptions/subscribing-now:acknowledge)
    ----------------------------------------------------------------------
    ```

    From:

    1. https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/75334450
    1. https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/77188348
    1. https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/77564743
    1. https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/94920686

1.  Only one message returned

    ```python
    ======================================================================
    ERROR: test_message_pull_mode_e2e (pubsub.TestPubsub)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/travis/build/GoogleCloudPlatform/gcloud-python/system_tests/pubsub.py", line 129, in test_message_pull_mode_e2e
        message1, message2 = sorted(messages, key=_by_timestamp)
    ValueError: need more than 1 value to unpack
    ----------------------------------------------------------------------
    ```

    From: 

    1.  https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/77052107
    1.  https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/98407178

1.  Eventual Consistency Error

    ```python
    ======================================================================
    FAIL: test_list_subscriptions (pubsub.TestPubsub)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/travis/build/GoogleCloudPlatform/gcloud-python/system_tests/pubsub.py", line 100, in test_list_subscriptions
        self.assertEqual(len(created), len(subscriptions_to_create))
    AssertionError: 2 != 3
    ======================================================================
    FAIL: test_list_topics (pubsub.TestPubsub)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/home/travis/build/GoogleCloudPlatform/gcloud-python/system_tests/pubsub.py", line 62, in test_list_topics
        self.assertEqual(len(created), len(topics_to_create))
    AssertionError: 2 != 3
    ----------------------------------------------------------------------
    Ran 5 tests in 66.301s
    FAILED (failures=2)
    ```

    From:

    1.  https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/77600321
    1.  https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/78109143

**Comment by dhermes** [comment]
Same issue in https://travis-ci.org/GoogleCloudPlatform/gcloud-python/builds/75334450 (build was from #1060, previously reported in #1062)

**Comment by dhermes** [comment]
@tseaver @tmatsuo The 1st error occurred again today.

@tmatsuo Not sure if there is a place to report this internally, but we are hitting this with a tiny amount of load and we've see this server-side flakiness 4 times now. 99% of the time it works (i.e. it knows there is an Ack ID in the request) but for some reason it doesn't always pick it up.

**Comment by tmatsuo** [comment]
@dhermes 
99%, I think it matches the end-to-end message delivery latency expectations at this point.

The tail latency is mainly caused by server crashes, disk crashes, or network cut etc (but we never lost any messages so far, that's great), as you can imagine, and it's almost inevitable.

Only thing I can think of is longer wait time (and more pull calls) for the tests. Is it possible to have a time limit, and pulls again and again until it reaches the time limit?

**Comment by dhermes** [comment]
Ah I see. So in the [source][1]

```python
ack_ids = [recv[0] for recv in received]
subscription.acknowledge(ack_ids)
```

we should check that `received` / `ack_ids` is not empty and then never send the acknowledge request?

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/02a304065460435eb1447e61938945865b2c0995/system_tests/pubsub.py#L138-L139

**Comment by tmatsuo** [comment]
The response can be contain message, 1 message, or 2 messages. You can repeatedly call pull until you get the right amount of messages. sg?

**Comment by dhermes** [comment]
> can be contain message

You mean 0 messages?

**Comment by dhermes** [comment]
Also 

> repeatedly call pull until you get the right amount of messages

does sound good

**Comment by tmatsuo** [comment]
@dhermes 
> You mean 0 messages?
Yes, I corrected my comment :)

**Comment by dhermes** [comment]
:+1:

**Comment by daspecster** [comment]
@dhermes, for point 4, I tried playing with `@Retry` in those methods and retrying didn't seem to work.

```bash
test_create_subscription_defaults (pubsub.TestPubsub) ... ok
test_create_subscription_w_ack_deadline (pubsub.TestPubsub) ... ok
test_create_topic (pubsub.TestPubsub) ... ok
test_fetch_delete_subscription_w_deleted_topic (pubsub.TestPubsub) ... 0 != 1, Trying again in 30 seconds...
0 != 1, Trying again in 60 seconds...

0 != 1, Trying again in 120 seconds...
FAIL
test_list_subscriptions (pubsub.TestPubsub) ... 2 != 3, Trying again in 30 seconds...
2 != 3, Trying again in 60 seconds...
2 != 3, Trying again in 120 seconds...
2 != 3, Trying again in 30 seconds...
True is not false, Trying again in 60 seconds...
True is not false, Trying again in 120 seconds...
FAIL
test_list_topics (pubsub.TestPubsub) ... 2 != 3, Trying again in 30 seconds...
2 != 3, Trying again in 60 seconds...
2 != 3, Trying again in 120 seconds...
FAIL
test_message_pull_mode_e2e (pubsub.TestPubsub) ... ok
test_subscription_iam_policy (pubsub.TestPubsub) ... ok
test_topic_iam_policy (pubsub.TestPubsub) ... ok
```

**Comment by dhermes** [comment]
@daspecster When you say you tried, did you decorate the entire test case or just the sensitive part?

**Comment by tseaver** [comment]
#2050 fixes the "#4. Eventual consistency errors" case for me.

**Comment by daspecster** [comment]
@dhermes just the asserts actually.

@tseaver fantastic! I'm excited for master to turn green again.

**Comment by tseaver** [comment]
Point 3) now tracked in #2077.  Point 1) is really the same issue, with the failure in parsing the empty response already fixed in #899.  Point 2) is also a symptom of #2077.
